### ISSUE TITLE
🤗 [Question]: Guidance Needed on Usage of c.Locals(), c.Context(), and c.UserContext()

### ISSUE THREAD

**Issue by gopkg-dev** [gofiber/fiber#2717]
### Question Description

Hello,

I am working on a GoFiber project and I am encountering some confusion regarding the usage of `c.Locals()`, `c.Context()`, and `c.UserContext()`. 

I understand that:

1. `c.Locals()` is used to pass data within the current request cycle.
2. `c.Context()` provides access to the native `net/http` context.
3. `c.UserContext()` is a feature from `fasthttp` that allows storing data in `fasthttp.RequestCtx`.

In the above code, I set a value in `c.Locals()` and attempt to retrieve it using `c.Context().Value()`. However, I am unsure if this is the correct way to handle this situation. 

Could you please provide some guidance on when to use `c.Locals()`, `c.Context()`, and `c.UserContext()` in such scenarios?

Thank you for your assistance.

### Code Snippet (optional)

```go
package main

import (
	"fmt"
	"log"

	"github.com/gofiber/fiber/v2"
)

func main() {
	app := fiber.New()

	type userIDCtx struct{}

	// A middleware to set value
	app.Use(func(c *fiber.Ctx) error {
		c.Locals(userIDCtx{}, "1")

		//c.SetUserContext(context.WithValue(c.UserContext(),userIDCtx{}, "1"))
		return c.Next()
	})

	// An example to describe the question
	app.Get("/", func(c *fiber.Ctx) error {
		fmt.Println(c.Context().Value(userIDCtx{}).(string)) // print 1
		fmt.Println(c.Locals(userIDCtx{}).(string))          // print 1

		//fmt.Println(c.UserContext().Value(userIDCtx{}).(string))

		//c.UserContext()
		//c.Context()
		//c.Locals()

		return c.SendStatus(fiber.StatusOK)
	})
	log.Fatal(app.Listen(":3000"))
}
```

### Checklist:

- [X] I agree to follow Fiber's [Code of Conduct](https://github.com/gofiber/fiber/blob/master/.github/CODE_OF_CONDUCT.md).
- [X] I have checked for existing issues that describe my questions prior to opening this one.
- [X] I understand that improperly formatted questions may be closed without explanation.

**Comment by ReneWerner87** [comment]
I can clarify the usage of `c.Locals()`, `c.Context()`, and `c.UserContext()` for you.

1. `c.Locals()` is used to pass data within the current request cycle, storing data in the context of the `fiber.Ctx` instance.

2. `c.Context()` provides access to the native `fasthttp.RequestCtx`, the underlying HTTP request context from Fasthttp, carrying information like the HTTP request and response.

3. `c.UserContext()` offers a `context.Context` which you can set or fetch. This context is not directly linked to Fasthttp but provides a way to transfer custom data and values across your application.

In your code snippet:

- When setting a value with `c.Locals()`, retrieve it using `c.Locals()` as well, since `c.Context().Value()` accesses the `fasthttp.RequestCtx`, which is different from `fiber.Ctx` locals.
- To correctly pass a `context.Context` type to your service layer and retrieve values set in middleware, use `c.UserContext()`. This method allows you to transfer a custom context through the middleware to your service handlers.

In summary, use `c.Locals()` for values relevant within the current request cycle, `c.Context()` for accessing Fasthttp-specific functionalities, and `c.UserContext()` for transferring context information through middleware in your service layer.

**Comment by suntong** [comment]
That's exactly the question that I have in mind. 

Thanks @gopkg-dev & @ReneWerner87 !

For a further follow up, 

> I am particularly interested in understanding how to correctly pass a context.Context type to my service layer and retrieve the values set in the middleware.

What I want to achieve, is somewhat like AWS's context tracing, i.e., if a request hit the web server, we can trace that specific request with a specific request id that's unique to that web request, so as to tigh all backend services relevant/servicing to that specific web request, all the way to writing to DB.

I guess it involves generating such specific request id when a request hit the web server, passing it all along the services trails, and eventually accessing DB with it, along with `context.WithTimeout` like [here](https://www.mongodb.com/community/forums/t/golang-driver-context/6389). 

Would anyone give a sample code highlight the above key points please? thx a lot!

**Comment by ReneWerner87** [comment]
why not just using https://docs.gofiber.io/api/middleware/requestid ?
You can then access this value via locals

**Comment by suntong** [comment]
Well, I didn't know that thing exist until you mention it. Thanks.

**Comment by suntong** [comment]
> You can then access this value via **locals**

> ContextKey defines the key used when storing the request ID in the **locals** for a specific request.

Note to myself, here the "**_locals_**" means, 
https://docs.gofiber.io/api/ctx/#locals

**Comment by lynx-check-in** [comment]
So, if I used the JWT middleware that stores the parsed JWT in the c.Locals(), can I then store the `sub` (for example) of that JWT in the c.UserContext() and pass that context from the handler to a function to call the DB and have that function retrieve the `sub` from the context safely, knowing that those values are specific to the current request cycle?

```go
func AuthMiddleware() fiber.Handler {
    return func(c *fiber.Ctx) error {
        ctx := c.UserContext()
        
        // Extract claims
        claims := c.Locals("user").(*jwt.Token).Claims.(jwt.MapClaims)
        sub := claims["sub"].(string)
        accountID := claims["custom:account"].(string)
        
        // Add our values
        ctx = context.WithValue(ctx, contextkeys.SubjectKey, sub)
        
        // Update
        c.SetUserContext(ctx)
        
        return c.Next()
    }
}

func GetUserHandler(c *fiber.Ctx) error {
    ctx := c.UserContext()
    userID := c.Params("id")

    user, _ := GetUserDB(ctx, userID)

    return c.JSON(user)
}

func GetUserDB(ctx context.Context, userID string) User, error {
    sub := ctx.Value(contextkeys.SubjectKey).(string)
    ...
}
```

Would this be safe, or should we pass the `sub` from the middleware to the handler using c.Locals()?

**Comment by ReneWerner87** [comment]
Both is possible and safe.
I prefer the locals in the handler scope and the context.context value in every scope which is outside in an independent scope.
You example is good and right.

**Comment by gaby** [comment]
Explained in https://github.com/gofiber/fiber/issues/2717#issuecomment-1805472836
